src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,450 @@
+                    Ordering::SeqCst,
+                );
+                unsafe {
+                    cell.store::<Address>(last);
+                }
+                last = cell;
+            }
+            cell += cell_size;
+        }
+
+        self.store_free_list(last);
+    }
+
+    /// This is a naive implementation that is inefficient but should be correct.
+    /// In this implementation, we simply go through each possible object
+    /// reference and see if it has the mark bit set. If we find mark bit, that means the cell is alive. If we didn't find
+    /// the mark bit in the entire cell, it means the cell is dead.
+    fn naive_brute_force_sweep<VM: VMBinding>(&self) {
+        use crate::util::constants::MIN_OBJECT_SIZE;
+
+        // Cell size for this block.
+        let cell_size = self.load_block_cell_size();
+        // Current cell
+        let mut cell = self.start();
+        // Last free cell in the free list
+        let mut last = Address::ZERO;
+        // Current cursor
+        let mut cursor = cell;
+
+        debug!("Sweep block {:?}, cell size {}", self, cell_size);
+
+        while cell + cell_size <= self.end() {
+            // possible object ref
+            let potential_object_ref = ObjectReference::from_raw_address(
+                cursor + VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND,
+            );
+            trace!(
+                "{:?}: cell = {}, last cell in free list = {}, cursor = {}, potential object = {}",
+                self,
+                cell,
+                last,
+                cursor,
+                potential_object_ref
+            );
+
+            if VM::VMObjectModel::LOCAL_MARK_BIT_SPEC
+                .is_marked::<VM>(potential_object_ref, Ordering::SeqCst)
+            {
+                debug!("{:?} Live cell: {}", self, cell);
+                // If the mark bit is set, the cell is alive.
+                // We directly jump to the end of the cell.
+                cell += cell_size;
+                cursor = cell;
+            } else {
+                // If the mark bit is not set, we don't know if the cell is alive or not. We keep search for the mark bit.
+                cursor += MIN_OBJECT_SIZE;
+
+                if cursor >= cell + cell_size {
+                    // We now stepped to the next cell. This means we did not find mark bit in the current cell, and we can add this cell to free list.
+                    debug!(
+                        "{:?} Free cell: {}, last cell in freelist is {}",
+                        self, cell, last
+                    );
+
+                    // Clear alloc bit: we don't know where the object reference actually is, so we bulk zero the possible area.
+                    #[cfg(feature = "global_alloc_bit")]
+                    crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.bzero_metadata(
+                        cell + VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND,
+                        cell_size,
```

Comment:
If we know that an object cannot span multiple cells, we don't need to clear this many bytes.  We only need to clear the metadata bits for all bytes from `cell + VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND` to the end of this cell.

Reply:
An object will not span multiple cells. But I don't quite understand what you said. This code uses bulk zero to clear one cell, as we do not know where the object is, and where the alloc bit is set.

Reply:
For example, if the cell size is 128 bytes, and `OBJECT_REF_OFFSET_LOWER_BOUND` is equal to 24, then we only need to clear the metadata bits for the interval of `[cell + 24, cell + 128)` instead of `[cell + 24, cell + 24 + 128)`.  The VO-bit cannot be set in the range of `[cell + 128, cell + 128 + 24)` because that is in the next cell.

Concretely,
```
crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.bzero_metadata(
                        cell + VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND,
                        cell_size - VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND,
```

Reply:
I changed to zero from `cell` for `cell_size` bytes. As we now guarantee we mark object metadata within its allocation (i.e. a cell here), we do not really need `OBJECT_REF_OFFSET_LOWER_BOUND`.